cli/central-client/src/test/java/org/ballerinalang/central/client/TestCentralApiClient.java

```diff
@@ -208,6 +227,93 @@ public void testPullNonExistingPackage() throws IOException, CentralClientExcept
         this.pullPackage("foo", "sf", "1.3.5", TMP_DIR, ANY_PLATFORM, TEST_BAL_VERSION, false);
     }
 
+    @DataProvider (name = "provideDataForBalaDownloadFailHttpCodes")
+    private Object[][] provideDataForBalaDownloadFailHttpCodes() {
+
+        // invalidRequestErrorMessage here might not be identical to the actual HTTP response error message
+        String invalidResponseErrorMessage = "package not found: foo/sf:*_any";
+        String errorWithNoResponseBodyMessage = "error: failed to pull the package: 'foo/sf:1.3.5'. " +
+                "BALA content download from '" + this.balaUrl + "' failed.";
+        String remoteRepositoryErrorMessage = errorWithNoResponseBodyMessage + " reason:" + invalidResponseErrorMessage;
+        String responseString = "{\"message\": \"" + invalidResponseErrorMessage + "\"}";
+
+        Object[][] data = new Object[6][3];
+        data[0] = new Object[]{HttpURLConnection.HTTP_NOT_FOUND, invalidResponseErrorMessage,
+                getResponseBody(responseString)};
+        data[1] = new Object[]{HttpURLConnection.HTTP_INTERNAL_ERROR, remoteRepositoryErrorMessage,
+                getResponseBody(responseString)};
+        data[2] = new Object[]{HttpURLConnection.HTTP_BAD_GATEWAY, remoteRepositoryErrorMessage,
+                getResponseBody(responseString)};
+        data[3] = new Object[]{HttpURLConnection.HTTP_UNAVAILABLE, remoteRepositoryErrorMessage,
+                getResponseBody(responseString)};
+        data[4] = new Object[]{HttpURLConnection.HTTP_GATEWAY_TIMEOUT, remoteRepositoryErrorMessage,
+                getResponseBody(responseString)};
+
+        // test BALA download with no response body
+        data[5] = new Object[]{HttpURLConnection.HTTP_UNAVAILABLE, errorWithNoResponseBodyMessage, null};
```

Comment:
Shall we add a test case to validate the behavior when an unexpected error code is received such as `HTTP_FORBIDDEN`?